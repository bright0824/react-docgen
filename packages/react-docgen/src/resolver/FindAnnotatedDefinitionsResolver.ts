import normalizeClassDefinition from '../utils/normalizeClassDefinition.js';
import type { NodePath } from '@babel/traverse';
import { visitors } from '@babel/traverse';
import type FileState from '../FileState.js';
import type { ComponentNodePath, ResolverClass } from './index.js';
import type {
  ArrowFunctionExpression,
  ClassDeclaration,
  ClassExpression,
  FunctionDeclaration,
  FunctionExpression,
  ObjectMethod,
} from '@babel/types';

interface TraverseState {
  foundDefinitions: Set<ComponentNodePath>;
  annotation: string;
}

function isAnnotated(path: NodePath, annotation: string): boolean {
  let inspectPath: NodePath | null = path;

  do {
    const leadingComments = inspectPath.node.leadingComments;

    // If an export doesn't have leading comments, we can simply continue
    if (leadingComments && leadingComments.length > 0) {
      // Search for the annotation in any comment.
      const hasAnnotation = leadingComments.some(({ value }) =>
        value.includes(annotation),
      );

      // if we found an annotation return true
      if (hasAnnotation) {
        return true;
      }
    }

    // return false if the container of the current path is an array
    // as we do not want to traverse up through this kind of nodes, like ArrayExpressions for example
    // The only exception is variable declarations
    if (
      Array.isArray(inspectPath.container) &&
      !inspectPath.isVariableDeclarator() &&
      !inspectPath.parentPath?.isCallExpression()
    ) {
      return false;
    }
  } while ((inspectPath = inspectPath.parentPath));

  return false;
}

function classVisitor(
  path: NodePath<ClassDeclaration | ClassExpression>,
  state: TraverseState,
): void {
  if (isAnnotated(path, state.annotation)) {
    normalizeClassDefinition(path);
    state.foundDefinitions.add(path);
  }
}

function statelessVisitor(
  path: NodePath<
    | ArrowFunctionExpression
    | FunctionDeclaration
    | FunctionExpression
    | ObjectMethod
  >,
  state: TraverseState,
): void {
  if (isAnnotated(path, state.annotation)) {
    state.foundDefinitions.add(path);
  }
}

const explodedVisitors = visitors.explode<TraverseState>({
  ArrowFunctionExpression: { enter: statelessVisitor },
  FunctionDeclaration: { enter: statelessVisitor },
  FunctionExpression: { enter: statelessVisitor },
  ObjectMethod: { enter: statelessVisitor },

  ClassDeclaration: { enter: classVisitor },
  ClassExpression: { enter: classVisitor },
});

interface FindAnnotatedDefinitionsResolverOptions {
  annotation?: string;
}

/**
 * Given an AST, this function tries to find all react components which
 * are annotated with an annotation
 */
export default class FindAnnotatedDefinitionsResolver implements ResolverClass {
  annotation: string;

  constructor({
    annotation = '@component',
  }: FindAnnotatedDefinitionsResolverOptions = {}) {
    this.annotation = annotation;
  }

  resolve(file: FileState): ComponentNodePath[] {
    const state: TraverseState = {
      foundDefinitions: new Set<ComponentNodePath>(),
      annotation: this.annotation,
    };

    file.traverse(explodedVisitors, state);

    return Array.from(state.foundDefinitions);
  }
}
