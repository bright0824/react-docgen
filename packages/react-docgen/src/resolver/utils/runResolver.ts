import type FileState from '../../FileState.js';
import type { Resolver, ResolverClass, ResolverFunction } from '../index.js';

function isResolverClass(resolver: Resolver): resolver is ResolverClass {
  return typeof resolver === 'object';
}

export default function runResolver(
  resolver: Resolver,
  file: FileState,
): ReturnType<ResolverFunction> {
  return isResolverClass(resolver) ? resolver.resolve(file) : resolver(file);
}
