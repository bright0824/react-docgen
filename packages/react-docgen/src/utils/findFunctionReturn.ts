import type { NodePath } from '@babel/traverse';
import { visitors } from '@babel/traverse';
import resolveToValue from './resolveToValue';
import { ignore } from './traverse';

type Predicate = (p: NodePath) => boolean;
interface TraverseState {
  readonly predicate: Predicate;
  resolvedReturnPath?: NodePath;
  readonly seen: WeakSet<NodePath>;
}

const explodedVisitors = visitors.explode<TraverseState>({
  Function: { enter: ignore },
  Class: { enter: ignore },
  ObjectExpression: { enter: ignore },
  ReturnStatement: {
    enter: function (path, state) {
      const argument = path.get('argument');

      if (argument.hasNode()) {
        const resolvedPath = resolvesToFinalValue(
          argument,
          state.predicate,
          state.seen,
        );

        if (resolvedPath) {
          state.resolvedReturnPath = resolvedPath;
          path.stop();
        }
      }
    },
  },
});

function resolvesToFinalValue(
  path: NodePath,
  predicate: Predicate,
  seen: WeakSet<NodePath>,
): NodePath | undefined {
  // avoid returns with recursive function calls
  if (seen.has(path)) {
    return;
  }
  seen.add(path);

  // Is the path already passes then return it.
  if (predicate(path)) {
    return path;
  }

  const resolvedPath = resolveToValue(path);

  // If the resolved path is already passing then no need to further check
  // Only do this if the resolvedPath actually resolved something as otherwise we did this check already
  if (resolvedPath.node !== path.node && predicate(resolvedPath)) {
    return resolvedPath;
  }

  // If the path points to a conditional expression, then we need to look only at
  // the two possible paths
  if (resolvedPath.isConditionalExpression()) {
    return (
      resolvesToFinalValue(resolvedPath.get('consequent'), predicate, seen) ||
      resolvesToFinalValue(resolvedPath.get('alternate'), predicate, seen)
    );
  }

  // If the path points to a logical expression (AND, OR, ...), then we need to look only at
  // the two possible paths
  if (resolvedPath.isLogicalExpression()) {
    return (
      resolvesToFinalValue(resolvedPath.get('left'), predicate, seen) ||
      resolvesToFinalValue(resolvedPath.get('right'), predicate, seen)
    );
  }

  // If we have a call expression, lets try to follow it
  if (resolvedPath.isCallExpression()) {
    const returnValue = findFunctionReturnWithCache(
      resolveToValue(resolvedPath.get('callee')),
      predicate,
      seen,
    );

    if (returnValue) {
      return returnValue;
    }
  }

  return;
}

/**
 * This can be used in two ways
 * 1. Find the first return path that passes the predicate function
 *    (for example to check if a function is returning something)
 * 2. Find all occurrences of return values
 *    For this the predicate acts more like a collector and always needs to return false
 */
function findFunctionReturnWithCache(
  path: NodePath,
  predicate: Predicate,
  seen: WeakSet<NodePath>,
): NodePath | undefined {
  let functionPath: NodePath = path;

  if (functionPath.isObjectProperty()) {
    functionPath = functionPath.get('value');
  } else if (functionPath.isClassProperty()) {
    const classPropertyValue = functionPath.get('value');

    if (classPropertyValue.hasNode()) {
      functionPath = classPropertyValue;
    }
  }

  if (!functionPath.isFunction()) {
    return;
  }

  // skip traversing for ArrowFunctionExpressions with no block
  if (path.isArrowFunctionExpression()) {
    const body = path.get('body');

    if (!body.isBlockStatement()) {
      return resolvesToFinalValue(body, predicate, seen);
    }
  }

  const state: TraverseState = {
    predicate,
    seen,
  };

  path.traverse(explodedVisitors, state);

  return state.resolvedReturnPath;
}

/**
 * This can be used in two ways
 * 1. Find the first return path that passes the predicate function
 *    (for example to check if a function is returning something)
 * 2. Find all occurrences of return values
 *    For this the predicate acts more like a collector and always needs to return false
 */
export default function findFunctionReturn(
  path: NodePath,
  predicate: Predicate,
): NodePath | undefined {
  return findFunctionReturnWithCache(path, predicate, new WeakSet());
}
