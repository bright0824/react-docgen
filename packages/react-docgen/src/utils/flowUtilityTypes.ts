import type { NodePath } from '@babel/traverse';
import type { GenericTypeAnnotation } from '@babel/types';

/**
 * See `supportedUtilityTypes` for which types are supported and
 * https://flow.org/en/docs/types/utilities/ for which types are available.
 */
export function isSupportedUtilityType(
  path: NodePath,
): path is NodePath<GenericTypeAnnotation> {
  if (path.isGenericTypeAnnotation()) {
    const idPath = path.get('id');

    if (idPath.isIdentifier()) {
      const name = idPath.node.name;

      return name === '$Exact' || name === '$ReadOnly';
    }
  }

  return false;
}

/**
 * Unwraps well known utility types. For example:
 *
 *   $ReadOnly<T> => T
 */
export function unwrapUtilityType(path: NodePath): NodePath {
  let resultPath: NodePath = path;
  while (isSupportedUtilityType(resultPath)) {
    const typeParameters = resultPath.get('typeParameters');
    if (!typeParameters.hasNode()) break;

    const firstParam = typeParameters.get('params')[0];
    if (!firstParam) break;

    resultPath = firstParam;
  }

  return resultPath;
}
