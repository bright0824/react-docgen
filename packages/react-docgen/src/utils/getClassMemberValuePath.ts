import type { NodePath } from '@babel/traverse';
import type {
  ClassDeclaration,
  ClassExpression,
  ClassMethod,
  ClassProperty,
  Expression,
} from '@babel/types';
import getNameOrValue from './getNameOrValue';

export default function getClassMemberValuePath(
  classDefinition: NodePath<ClassDeclaration | ClassExpression>,
  memberName: string,
): NodePath<ClassMethod | Expression> | null {
  const classMember = classDefinition
    .get('body')
    .get('body')
    .find(memberPath => {
      if (
        (memberPath.isClassMethod() && memberPath.node.kind !== 'set') ||
        memberPath.isClassProperty()
      ) {
        const key = (memberPath as NodePath<ClassMethod | ClassProperty>).get(
          'key',
        );

        return (
          (!memberPath.node.computed || key.isLiteral()) &&
          getNameOrValue(key) === memberName
        );
      }

      return false;
    });

  if (classMember) {
    // For ClassProperty we return the value and for ClassMethod
    // we return itself
    return classMember.isClassMethod()
      ? classMember
      : (classMember.get('value') as NodePath<Expression>);
  }

  return null;
}
