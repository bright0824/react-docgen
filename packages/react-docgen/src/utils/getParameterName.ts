import type { NodePath } from '@babel/traverse';
import type {
  ArrayPattern,
  AssignmentPattern,
  Identifier,
  ObjectPattern,
  RestElement,
  TSParameterProperty,
} from '@babel/types';
import printValue from './printValue.js';

type ParameterNodePath = NodePath<
  | ArrayPattern
  | AssignmentPattern
  | Identifier
  | ObjectPattern
  | RestElement
  | TSParameterProperty
>;

export default function getParameterName(
  parameterPath: ParameterNodePath,
): string {
  if (parameterPath.isIdentifier()) {
    return parameterPath.node.name;
  } else if (parameterPath.isAssignmentPattern()) {
    return getParameterName(parameterPath.get('left') as ParameterNodePath);
  } else if (
    parameterPath.isObjectPattern() ||
    parameterPath.isArrayPattern()
  ) {
    return printValue(parameterPath);
  } else if (parameterPath.isRestElement()) {
    return `...${getParameterName(
      parameterPath.get('argument') as ParameterNodePath,
    )}`;
  } else if (parameterPath.isTSParameterProperty()) {
    return getParameterName(parameterPath.get('parameter'));
  }

  throw new TypeError(
    'Parameter name must be one of Identifier, AssignmentPattern, ArrayPattern, ' +
      `ObjectPattern or RestElement, instead got ${parameterPath.node.type}`,
  );
}
