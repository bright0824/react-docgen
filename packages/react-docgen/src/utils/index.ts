import * as docblock from './docblock';
import * as expressionTo from './expressionTo';
import * as flowUtilityTypes from './flowUtilityTypes';
import * as traverse from './traverse';

export { docblock, expressionTo, flowUtilityTypes, traverse };

export { default as getClassMemberValuePath } from './getClassMemberValuePath';
export { default as getFlowType } from './getFlowType';
export { default as getMemberExpressionRoot } from './getMemberExpressionRoot';
export { default as getMemberExpressionValuePath } from './getMemberExpressionValuePath';
export { default as getMembers } from './getMembers';
export {
  default as getMemberValuePath,
  isSupportedDefinitionType,
} from './getMemberValuePath';
export { default as getMethodDocumentation } from './getMethodDocumentation';
export type { MethodNodePath } from './getMethodDocumentation';
export { default as getNameOrValue } from './getNameOrValue';
export { default as getParameterName } from './getParameterName';
export { default as getPropertyName, COMPUTED_PREFIX } from './getPropertyName';
export { default as getPropertyValuePath } from './getPropertyValuePath';
export { default as getPropType } from './getPropType';
export { default as getTSType } from './getTSType';
export { default as getTypeAnnotation } from './getTypeAnnotation';
export {
  default as getTypeFromReactComponent,
  applyToTypeProperties,
} from './getTypeFromReactComponent';
export { default as getTypeIdentifier } from './getTypeIdentifier';
export { default as getTypeParameters } from './getTypeParameters';
export type { TypeParameters } from './getTypeParameters';
export { default as isDestructuringAssignment } from './isDestructuringAssignment';
export { default as isExportsOrModuleAssignment } from './isExportsOrModuleAssignment';
export { default as isReactBuiltinCall } from './isReactBuiltinCall';
export { default as isReactChildrenElementCall } from './isReactChildrenElementCall';
export { default as isReactCloneElementCall } from './isReactCloneElementCall';
export { default as isReactComponentClass } from './isReactComponentClass';
export { default as isReactComponentMethod } from './isReactComponentMethod';
export { default as isReactCreateClassCall } from './isReactCreateClassCall';
export { default as isReactCreateElementCall } from './isReactCreateElementCall';
export { default as isReactForwardRefCall } from './isReactForwardRefCall';
export { default as isReactModuleName } from './isReactModuleName';
export { default as isRequiredPropType } from './isRequiredPropType';
export { default as isStatelessComponent } from './isStatelessComponent';
export { default as isUnreachableFlowType } from './isUnreachableFlowType';
export { default as match } from './match';
export { default as normalizeClassDefinition } from './normalizeClassDefinition';
export { default as parseJsDoc } from './parseJsDoc';
export { default as postProcessDocumentation } from './postProcessDocumentation';
export { default as printValue } from './printValue';
export { default as resolveExportDeclaration } from './resolveExportDeclaration';
export { default as resolveFunctionDefinitionToReturnValue } from './resolveFunctionDefinitionToReturnValue';
export { default as resolveGenericTypeAnnotation } from './resolveGenericTypeAnnotation';
export { default as resolveHOC } from './resolveHOC';
export { default as resolveObjectPatternPropertyToValue } from './resolveObjectPatternPropertyToValue';
export {
  default as resolveObjectKeysToArray,
  resolveObjectToNameArray,
} from './resolveObjectKeysToArray';
export { default as resolveObjectValuesToArray } from './resolveObjectValuesToArray';
export { default as resolveToModule } from './resolveToModule';
export { default as resolveToValue } from './resolveToValue';
export { default as setPropDescription } from './setPropDescription';
