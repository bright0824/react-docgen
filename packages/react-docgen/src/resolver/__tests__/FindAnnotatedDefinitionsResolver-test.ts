import type { NodePath } from '@babel/traverse';
import { parse, noopImporter } from '../../../tests/utils';
import FindAnnotatedDefinitionsResolver from '../FindAnnotatedDefinitionsResolver.js';
import { describe, expect, test } from 'vitest';

describe('FindAnnotatedDefinitionsResolver', () => {
  const resolver = new FindAnnotatedDefinitionsResolver();

  function findComponentsInSource(
    source: string,
    importer = noopImporter,
  ): NodePath[] {
    return resolver.resolve(parse(source, {}, importer, true));
  }

  describe('class definitions', () => {
    test('finds ClassDeclaration with line comment', () => {
      const source = `
        // @component
        class Component {}
      `;

      expect(findComponentsInSource(source)).toMatchSnapshot();
    });

    test('finds ClassDeclaration with line comment and empty line', () => {
      const source = `
        // @component

        class Component {}
      `;

      expect(findComponentsInSource(source)).toMatchSnapshot();
    });

    test('finds ClassDeclaration with block comment', () => {
      const source = `
        /* @component */
        class Component {}
      `;

      expect(findComponentsInSource(source)).toMatchSnapshot();
    });

    test('finds ClassDeclaration with block comment and empty line', () => {
      const source = `
        /* @component */

        class Component {}
      `;

      expect(findComponentsInSource(source)).toMatchSnapshot();
    });

    test('finds ClassExpression', () => {
      const source = `
        // @component
        const Component = class {}
      `;

      expect(findComponentsInSource(source)).toMatchSnapshot();
    });

    test('finds ClassExpression with assignment', () => {
      const source = `
        let Component;
        // @component
        Component = class {}
      `;

      expect(findComponentsInSource(source)).toMatchSnapshot();
    });

    test('finds nothing when not annotated', () => {
      const source = `
        class ComponentA {}
        const ComponentB = class {}
      `;

      const result = findComponentsInSource(source);

      expect(Array.isArray(result)).toBe(true);
      expect(result.length).toBe(0);
    });
  });

  describe('stateless components', () => {
    test('finds ArrowFunctionExpression', () => {
      const source = `
        // @component
        const Component = () => {};
      `;

      expect(findComponentsInSource(source)).toMatchSnapshot();
    });

    test('finds FunctionDeclaration', () => {
      const source = `
        // @component
        function Component() {}
      `;

      expect(findComponentsInSource(source)).toMatchSnapshot();
    });

    test('finds FunctionExpression', () => {
      const source = `
        // @component
        const Component = function() {};
      `;

      expect(findComponentsInSource(source)).toMatchSnapshot();
    });

    test('finds ObjectMethod', () => {
      const source = `
        const obj = {
          // @component
          component() {}
        };
      `;

      expect(findComponentsInSource(source)).toMatchSnapshot();
    });

    test('finds ObjectProperty', () => {
      const source = `
        const obj = {
          // @component
          component: function() {}
        };
      `;

      expect(findComponentsInSource(source)).toMatchSnapshot();
    });

    test('finds nothing when not annotated', () => {
      const source = `
        const ComponentA = () => {};
        function ComponentB() {}
        const ComponentC = function() {};
        const obj = { component() {} };
        const obj2 = { component: function() {} };
      `;

      const result = findComponentsInSource(source);

      expect(Array.isArray(result)).toBe(true);
      expect(result.length).toBe(0);
    });
  });

  test('finds component wrapped in HOC', () => {
    const source = `
        // @component
        const Component = React.memo(() => {});
      `;

    expect(findComponentsInSource(source)).toMatchSnapshot();
  });

  test('finds named exported component', () => {
    const source = `
        // @component
        export const Component = () => {};
      `;

    expect(findComponentsInSource(source)).toMatchSnapshot();
  });

  test('finds default exported component', () => {
    const source = `
        // @component
        export default () => {};
      `;

    expect(findComponentsInSource(source)).toMatchSnapshot();
  });

  test('finds component wrapped in two HOCs', () => {
    const source = `
        // @component
        const Component = React.memo(otherHOC(() => {}));
      `;

    expect(findComponentsInSource(source)).toMatchSnapshot();
  });

  test('finds component wrapped in function', () => {
    const source = `
        function X () {
          // @component
          const subcomponent = class {}
        }
      `;

    expect(findComponentsInSource(source)).toMatchSnapshot();
  });

  test('does not traverse up ArrayExpressions', () => {
    const source = `
        // @component
        const arr = [
          function() {},
          function() {}
        ]
      `;

    const result = findComponentsInSource(source);

    expect(Array.isArray(result)).toBe(true);
    expect(result.length).toBe(0);
  });

  test('does not traverse up function parameters', () => {
    const source = `
        // @component
        function x(component = () => {}) {}
      `;

    const result = findComponentsInSource(source);

    expect(Array.isArray(result)).toBe(true);
    expect(result.length).toBe(1);
    expect(result[0].type).not.toBe('ArrowFunctionExpression');
  });

  test('finds function parameter with annotation', () => {
    const source = `
        function x(component = /* @component */() => {}) {}
      `;

    expect(findComponentsInSource(source)).toMatchSnapshot();
  });
});
