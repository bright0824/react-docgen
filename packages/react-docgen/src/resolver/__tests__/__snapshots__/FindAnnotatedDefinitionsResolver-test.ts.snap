// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FindAnnotatedDefinitionsResolver > class definitions > finds ClassDeclaration with block comment 1`] = `
[
  Node {
    "body": Node {
      "body": [],
      "type": "ClassBody",
    },
    "id": Node {
      "name": "Component",
      "type": "Identifier",
    },
    "superClass": null,
    "type": "ClassDeclaration",
  },
]
`;

exports[`FindAnnotatedDefinitionsResolver > class definitions > finds ClassDeclaration with block comment and empty line 1`] = `
[
  Node {
    "body": Node {
      "body": [],
      "type": "ClassBody",
    },
    "id": Node {
      "name": "Component",
      "type": "Identifier",
    },
    "superClass": null,
    "type": "ClassDeclaration",
  },
]
`;

exports[`FindAnnotatedDefinitionsResolver > class definitions > finds ClassDeclaration with line comment 1`] = `
[
  Node {
    "body": Node {
      "body": [],
      "type": "ClassBody",
    },
    "id": Node {
      "name": "Component",
      "type": "Identifier",
    },
    "superClass": null,
    "type": "ClassDeclaration",
  },
]
`;

exports[`FindAnnotatedDefinitionsResolver > class definitions > finds ClassDeclaration with line comment and empty line 1`] = `
[
  Node {
    "body": Node {
      "body": [],
      "type": "ClassBody",
    },
    "id": Node {
      "name": "Component",
      "type": "Identifier",
    },
    "superClass": null,
    "type": "ClassDeclaration",
  },
]
`;

exports[`FindAnnotatedDefinitionsResolver > class definitions > finds ClassExpression 1`] = `
[
  Node {
    "body": Node {
      "body": [],
      "type": "ClassBody",
    },
    "id": null,
    "superClass": null,
    "type": "ClassExpression",
  },
]
`;

exports[`FindAnnotatedDefinitionsResolver > class definitions > finds ClassExpression with assignment 1`] = `
[
  Node {
    "body": Node {
      "body": [],
      "type": "ClassBody",
    },
    "id": null,
    "superClass": null,
    "type": "ClassExpression",
  },
]
`;

exports[`FindAnnotatedDefinitionsResolver > finds component wrapped in HOC 1`] = `
[
  Node {
    "async": false,
    "body": Node {
      "body": [],
      "directives": [],
      "type": "BlockStatement",
    },
    "generator": false,
    "id": null,
    "params": [],
    "type": "ArrowFunctionExpression",
  },
]
`;

exports[`FindAnnotatedDefinitionsResolver > finds component wrapped in function 1`] = `
[
  Node {
    "body": Node {
      "body": [],
      "type": "ClassBody",
    },
    "id": null,
    "superClass": null,
    "type": "ClassExpression",
  },
]
`;

exports[`FindAnnotatedDefinitionsResolver > finds component wrapped in two HOCs 1`] = `
[
  Node {
    "async": false,
    "body": Node {
      "body": [],
      "directives": [],
      "type": "BlockStatement",
    },
    "generator": false,
    "id": null,
    "params": [],
    "type": "ArrowFunctionExpression",
  },
]
`;

exports[`FindAnnotatedDefinitionsResolver > finds function parameter with annotation 1`] = `
[
  Node {
    "async": false,
    "body": Node {
      "body": [],
      "directives": [],
      "type": "BlockStatement",
    },
    "generator": false,
    "id": null,
    "params": [],
    "type": "ArrowFunctionExpression",
  },
]
`;

exports[`FindAnnotatedDefinitionsResolver > stateless components > finds ArrowFunctionExpression 1`] = `
[
  Node {
    "async": false,
    "body": Node {
      "body": [],
      "directives": [],
      "type": "BlockStatement",
    },
    "generator": false,
    "id": null,
    "params": [],
    "type": "ArrowFunctionExpression",
  },
]
`;

exports[`FindAnnotatedDefinitionsResolver > stateless components > finds FunctionDeclaration 1`] = `
[
  Node {
    "async": false,
    "body": Node {
      "body": [],
      "directives": [],
      "type": "BlockStatement",
    },
    "generator": false,
    "id": Node {
      "name": "Component",
      "type": "Identifier",
    },
    "params": [],
    "type": "FunctionDeclaration",
  },
]
`;

exports[`FindAnnotatedDefinitionsResolver > stateless components > finds FunctionExpression 1`] = `
[
  Node {
    "async": false,
    "body": Node {
      "body": [],
      "directives": [],
      "type": "BlockStatement",
    },
    "generator": false,
    "id": null,
    "params": [],
    "type": "FunctionExpression",
  },
]
`;

exports[`FindAnnotatedDefinitionsResolver > stateless components > finds ObjectMethod 1`] = `
[
  Node {
    "async": false,
    "body": Node {
      "body": [],
      "directives": [],
      "type": "BlockStatement",
    },
    "computed": false,
    "generator": false,
    "id": null,
    "key": Node {
      "name": "component",
      "type": "Identifier",
    },
    "kind": "method",
    "method": true,
    "params": [],
    "type": "ObjectMethod",
  },
]
`;

exports[`FindAnnotatedDefinitionsResolver > stateless components > finds ObjectProperty 1`] = `
[
  Node {
    "async": false,
    "body": Node {
      "body": [],
      "directives": [],
      "type": "BlockStatement",
    },
    "generator": false,
    "id": null,
    "params": [],
    "type": "FunctionExpression",
  },
]
`;
